server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -1138,17 +1138,38 @@ public AliasFilter buildAliasFilter(ClusterState state, String index, Set<String
      */
     public CanMatchResponse canMatch(ShardSearchRequest request) throws IOException {
         assert request.searchType() == SearchType.QUERY_THEN_FETCH : "unexpected search type: " + request.searchType();
-        // TODO: support can_match with reader contexts after https://github.com/elastic/elasticsearch/pull/54966
-        assert request.readerId() == null : "request with reader_id bypass can_match phase";
         IndexService indexService = indicesService.indexServiceSafe(request.shardId().getIndex());
-        IndexShard indexShard = indexService.getShard(request.shardId().getId());
-        // we don't want to use the reader wrapper since it could run costly operations
-        // and we can afford false positives.
-        final boolean hasRefreshPending = indexShard.hasRefreshPending();
-        try (Engine.Searcher searcher = indexShard.acquireSearcher("can_match")) {
+        final boolean hasRefreshPending;
+        final Engine.Searcher canMatchSearcher;
+        if (request.readerId() != null) {
+            final ReaderContext readerContext = findReaderContext(request.readerId());
+            checkKeepAliveLimit(request.keepAlive().millis());
+            readerContext.keepAlive(request.keepAlive().millis());
+            readerContext.incRef();
```

Comment:
Why not using `markAsUsed` ?

Reply:
`markAsUsed` is not water-tight. It's possible that the Reaper already decides to reap this reader context, then `markAsUsed` will be a noop, and the searcher can be closed in midway (see https://github.com/elastic/elasticsearch/pull/20095). Should we make `markAsUsed` as tighten as `incRef`?